README

```diff
@@ -1451,41 +1451,173 @@ Normally for these values you can use:
 	node.conn[0].timeo.noop_out_interval = 5
 	node.conn[0].timeo.noop_out_timeout = 10
 
-If there are a lot of IO error messages, then the above values may be too
-aggressive and you may need to increase the values for your network conditions
-and workload, or you may need to check your network for possible problems.
+If there are a lot of IO error messages like
 
+detected conn error (22)
 
-8.1.2 replacement_timeout
-=========================
+in the kernel log then the above values may be too aggressive. You may need to
+increase the values for your network conditions and workload, or you may need
+to check your network for possible problems.
 
-The next iSCSI timer that will need to be tweaked is:
+8.1.2 SCSI command retries
+==========================
 
-node.session.timeo.replacement_timeout = X
+SCSI disk commands get 5 retries by default. In newer kernels this can be
+controlled via the sysfs file:
+
+/sys/block/$sdX/device/scsi_disk/$host:$bus:$target:LUN/max_retries
 
-Here X is in seconds.
+by writing a integer lower than 5 to reduce retries or setting to -1 for
+infinite retries.
 
-replacement_timeout will control how long to wait for session re-establishment
-before failing pending SCSI commands and commands that are being operated on by
-the SCSI layer's error handler up to a higher level like multipath or to
-an application if multipath is not being used.
+The number of actual retries a command gets may be less than 5 or what is
+requested in max_retries if the replacement timeout expires. When that timer
+expires it tells the SCSI layer to fail all new and queued commands.
 
 
-8.1.2.1 Running Commands, the SCSI Error Handler, and replacement_timeout
-=========================================================================
+8.1.3 replacement_timeout
+=========================
+
+The iSCSI layer timer:
 
-Remember from the Nop-out discussion that if a network problem is detected,
-the running commands are failed immediately. There is one exception to this,
-and that is when the SCSI layer's error handler is running. To check if
-the SCSI error handler is running, iscsiadm can be run as:
+node.session.timeo.replacement_timeout = X
+
+controls how long to wait for session re-establishment before failing all SCSI
+commands:
+
+1. commands that have been requeued and awaiting a retry
+2. commands that are being operated on by the SCSI layer's error handler
+3. all new commands that are queued to the device
+
+up to a higher level like multipath, filesystem layer, or to the application.
+
+The setting is in seconds. zero means to fail immediately. -1 means an infinite
+timeout which will wait until iscsid does a relogin, the user runs the iscsiadm
+logout command or until the node.session.reopen_max limit is hit.
+
+When this timer is started, the iSCSI layer will block new IO from executing
```

Comment:
Minor suggestion: use of the word "block" to indicate stopping/freezing new IO, and then referring to the "block" layer in the same sentence (and throughout the README) has potential for confusion. If it were possible to use a synonym for blocking that may prevent confusion in readers.

Of course this is standard usage for both of these terms, so I'm not claiming you're doing anything wrong here. It's just an unfortunate coincidence.

Reply:
I changed block layer to Block layer and block/blocking a device/queue to stop/stopping.